bindings/rust/bench/src/harness.rs

```diff
@@ -12,14 +19,145 @@ pub enum Mode {
     Server,
 }
 
-pub trait TlsBenchHarness {
+// these parameters were the only ones readily usable for all three libaries:
+// s2n-tls, rustls, and openssl
+#[allow(non_camel_case_types)]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum CipherSuite {
+    AES_128_GCM_SHA256,
+    AES_256_GCM_SHA384,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ECGroup {
+    SECP256R1,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct CryptoConfig {
+    pub cipher_suite: CipherSuite,
+    pub ec_group: ECGroup,
+}
+
+pub trait TlsBenchHarness: Sized {
+    /// Default harness
+    fn default() -> Result<Self, Box<dyn Error>> {
+        Self::new(&CryptoConfig {
+            cipher_suite: CipherSuite::AES_128_GCM_SHA256,
+            ec_group: ECGroup::SECP256R1,
+        })
+    }
+
     /// Initialize buffers, configs, and connections (pre-handshake)
-    fn new() -> Self;
+    fn new(crypto_config: &CryptoConfig) -> Result<Self, Box<dyn Error>>;
 
     /// Run handshake on initialized connection
-    /// Returns error if handshake already happened
-    fn handshake(&mut self) -> Result<(), &str>;
+    /// Returns error if handshake has already completed
+    fn handshake(&mut self) -> Result<(), Box<dyn Error>>;
 
     /// Checks if handshake is finished for both client and server
     fn handshake_completed(&self) -> bool;
+
+    /// Get negotiated cipher suite
+    ///
+    /// Returns `Err` if handshake has not completed
+    fn get_negotiated_cipher_suite(&self) -> Result<CipherSuite, Box<dyn Error>>;
+
+    /// Get whether or negotiated version is TLS1.3
+    ///
+    /// Retuns `Err` if handshake has not completed
+    fn negotiated_tls13(&self) -> Result<bool, Box<dyn Error>>;
+}
+
+/// Wrapper of two shared buffers to pass as stream
+/// This wrapper `read()`s into one buffer and `write()`s to another
+#[derive(Clone)]
+pub struct ConnectedBuffer {
+    recv: Rc<RefCell<VecDeque<u8>>>,
+    send: Rc<RefCell<VecDeque<u8>>>,
+}
+
+impl ConnectedBuffer {
```

Comment:
What is the point of this struct? It seems like you've already initialized the IO buffers in the individual harnesses.

Reply:
For Rustls and OpenSSL, you need a trait object implementing `Read + Write`, where you read and write messages to the other connection. Having `ConnectedBuffer` in `harness.rs` and not `rustls.rs` is mainly because OpenSSL would also need this trait impl. The IO buffers are also only ever initialized once, and we create new references to the buffers with `Rc::clone()`.

Reply:
I would probably not put this struct in this file, since it's just for benchmarking code. It would be better to put it in some sort of common.rs file that has shared code. If you don't want to do that you could just put this in the rustls.rs file for right now, since that's where it's being used.

Actually that probably goes for anything in this file, is there a reason why you defined the CipherSuites in this file too?

Reply:
I guess I'm cheating because I have extra context, but this struct is common to multiple benchmarks (or will be once we merge in the Openssl benchmark). If we want to could break it out into a utilities module? For the time being I'm okay with it living in the same file.

I think the CipherSuites are in this file because we need some way of abstracting over the individual configuration for each implementation, because individual implementations are so different (e.g. s2n-tls is the only one with a concept of security policies)

Reply:
I guess I've been treating harness.rs as a common file, since all other files already need to use harness anyways. I could separate the non-harness stuff into a common file.

Reply:
Since the harness would also be common to the s2n-tls and rustls impls, wouldn't the harness trait also belong in common.rs? I guess I've been treating harness.rs as common.rs, but I feel like having both might be a little redundant because both harness.rs and common.rs would have shared code.

Reply:
It's more like, there seems to be a distinction between the harness trait that each library has to implement in order to benchmark, and the code necessary to setup each library to handshake. So in my mind it makes the most sense to have those be in separate files. But it doesn't sound like anyone else thinks this distinction is important so I'm cool with approving it.